srml/kton/src/lib.rs

```diff
@@ -380,85 +377,29 @@ impl<T: Trait> Currency<T::AccountId> for Module<T> {
 
 impl<T: Trait> LockableCurrency<T::AccountId> for Module<T>
 where
-	T::Balance: MaybeSerializeDeserialize,
+	T::Balance: MaybeSerializeDeserialize + Debug,
 {
-	type Moment = T::BlockNumber;
+	type Id = Id<T::Moment>;
+	type WithdrawReasons = WithdrawReasons;
 
-	// `amount` > `free_balance` is allowed
-	fn set_lock(
-		id: LockIdentifier,
-		who: &T::AccountId,
-		amount: T::Balance,
-		until: T::BlockNumber,
-		reasons: WithdrawReasons,
-	) {
-		let now = <system::Module<T>>::block_number();
-		let mut new_lock = Some(BalanceLock {
-			id,
-			amount,
-			until,
-			reasons,
-		});
-		let mut locks = Self::locks(who)
-			.into_iter()
-			.filter_map(|l| {
-				if l.id == id {
-					new_lock.take()
-				} else if l.until > now {
-					Some(l)
-				} else {
-					None
-				}
-			})
-			.collect::<Vec<_>>();
-		if let Some(lock) = new_lock {
-			locks.push(lock);
-		}
-		<Locks<T>>::insert(who, locks);
-	}
+	fn set_lock(who: &T::AccountId, id: Self::Id, amount: Self::Balance, reasons: Self::WithdrawReasons) {
+		let now = <timestamp::Module<T>>::now();
 
-	fn extend_lock(
-		id: LockIdentifier,
-		who: &T::AccountId,
-		amount: T::Balance,
-		until: T::BlockNumber,
-		reasons: WithdrawReasons,
-	) {
-		let now = <system::Module<T>>::block_number();
-		let mut new_lock = Some(BalanceLock {
-			id,
-			amount,
-			until,
-			reasons,
+		<Locks<T>>::mutate(who, |locks| {
+			locks.retain(|lock| lock.id != id || lock.valid(&now));
+			locks.push(BalanceLock { id, amount, reasons });
```

Comment:
If there is a lock with same ID exists and also vaid, will it be retained with the new at the same time?

Reply:
I think this's very similar to the old logic, but with different ordering in `Vec`.

if exists:
- old logic: update the exists lock
- update: remove the exists lock, push the new lock